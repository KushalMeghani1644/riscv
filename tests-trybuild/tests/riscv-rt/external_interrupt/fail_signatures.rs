use riscv_rt::result::{Error, Result};

/// Just a dummy type to test the `ExternalInterrupt` trait.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum ExternalInterrupt {
    GPIO,
    UART,
    PWM,
}
unsafe impl riscv_rt::InterruptNumber for ExternalInterrupt {
    const MAX_INTERRUPT_NUMBER: usize = 2;

    #[inline]
    fn number(self) -> usize {
        self as usize
    }

    #[inline]
    fn from_number(value: usize) -> Result<Self> {
        match value {
            0 => Ok(Self::GPIO),
            1 => Ok(Self::UART),
            2 => Ok(Self::PWM),
            _ => Err(Error::InvalidVariant(value)),
        }
    }
}
unsafe impl riscv::ExternalInterruptNumber for ExternalInterrupt {}

#[riscv_rt::external_interrupt(ExternalInterrupt::GPIO)]
fn my_interrupt() -> usize {}

#[riscv_rt::external_interrupt(ExternalInterrupt::UART)]
fn my_other_interrupt(code: usize) -> usize {}

#[riscv_rt::external_interrupt(ExternalInterrupt::PWM)]
async fn my_async_interrupt(code: usize) -> usize {}

fn main() {}
