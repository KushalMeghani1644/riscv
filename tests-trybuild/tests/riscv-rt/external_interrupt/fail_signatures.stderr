error: `#[external_interrupt]` function must have signature `[unsafe] fn() [-> !]`
  --> tests/riscv-rt/external_interrupt/fail_signatures.rs:31:1
   |
31 | fn my_interrupt() -> usize {}
   | ^^

error: `#[external_interrupt]` function must have signature `[unsafe] fn() [-> !]`
  --> tests/riscv-rt/external_interrupt/fail_signatures.rs:34:1
   |
34 | fn my_other_interrupt(code: usize) -> usize {}
   | ^^

error: `#[external_interrupt]` function must have signature `[unsafe] fn() [-> !]`
  --> tests/riscv-rt/external_interrupt/fail_signatures.rs:37:1
   |
37 | async fn my_async_interrupt(code: usize) -> usize {}
   | ^^^^^
